File: WaveTest/ProductsTableViewController.m
#import "ProductsTableViewController.h"

#import "NetworkRequests.h"
#import "Product.h"

static NSString * const kWaveDeveloperAPI = @"https://api.waveapps.com/businesses/89746d57-c25f-4cec-9c63-34d7780b044b/products/";
static NSString * const kWaveDeveloperToken = @"d1hQ2z5dedmPGq8wVqU3vFLP4OstFh";


@interface ProductsTableViewController ()

@property (strong, nonatomic) NSMutableArray *waveProducts;

@end


@implementation ProductsTableViewController

#pragma mark - NSObject

- (void)viewDidLoad {
  [super viewDidLoad];
  
  self.waveProducts = [[NSMutableArray alloc] init];
  
  [self fetchProductsInBackground:self.waveProducts];
  [self configureUI];
}

#pragma mark - Configure UI

- (void)configureUI {
  
  [self configureNavBar];
}

- (void)configureNavBar {
  
  self.navigationItem.title = @"My Products";
}

#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
  return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
  
  if (self.waveProducts)
    return [self.waveProducts count];
  else
    return 0;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
  
  UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"productCell" forIndexPath:indexPath];
  [self configureCell:cell atIndexPath:indexPath];
  return cell;
}

#pragma mark - Table view helpers

- (void)configureCell:(UITableViewCell *)cell atIndexPath:(NSIndexPath *)indexPath {
  
  if (self.waveProducts) {
    Product *myProduct = [self.waveProducts objectAtIndex:indexPath.row];
    cell.textLabel.text = myProduct.name;
    cell.detailTextLabel.text = [NSString stringWithFormat:@"$%.2f", [myProduct.price doubleValue]];
  }
}

#pragma mark - Data

- (void)fetchProductsInBackground:(NSMutableArray *)myProducts {
  
  [NetworkRequests fetchDataInBackgroundFromStringURL:kWaveDeveloperAPI withAuthToken:kWaveDeveloperToken withCompletionBlock:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error) {
    
    // first, check for errors
    if (data == nil) {
      NSLog(@"Query failed!");
      return;
    }
    if (error) {
      NSLog(@"Error fetching product date. Description is %@", error);
      return;
    }
    NSError *jsonError = nil;
    NSArray *userData = [NSJSONSerialization JSONObjectWithData:data options:NSJSONReadingAllowFragments error:&jsonError];
    if (jsonError) {
      NSLog(@"JSON was malformed. Error is %@", jsonError);
      return;
    }
    
    // then, read and save data
    for (NSDictionary *product in userData) {
      NSString *name = [product objectForKey:@"name"];
      NSNumber *price = [NSNumber numberWithInt:[[product valueForKey:@"price"] doubleValue]];
      Product *newProduct = [[Product alloc] initWitName:name withPrice:price];
      [myProducts addObject:newProduct];
    }
    
    // finally, always update UI on main thread
    dispatch_async(dispatch_get_main_queue(), ^{
      [self.tableView reloadData];
    });
  }];
}

@end

File: wave-mobile-challenge-ios/wave-mobile-challenge-ios/ProductsTableViewController.swift
//
//  ProductsTableViewController.swift
//  wave-mobile-challenge-ios
//
//  Created by Naweed Tajuddin on 2018-04-03.
//  Copyright © 2018 Naweed Tajuddin. All rights reserved.
//

import UIKit

class ProductsTableViewController: UITableViewController
{

    var productsList = [Product]()
    let accessToken = "6W9hcvwRvyyZgPu9Odq7ko8DSY8Nfm"
    let productsEndpoint = "https://api.waveapps.com/businesses/89746d57-c25f-4cec-9c63-34d7780b044b/products/"
    
    override func viewDidLoad() {
        super.viewDidLoad()
    }
    
    override func viewWillAppear(_ animated: Bool) {
        
        self.navigationItem.title = "Billable Products"
        
        var urlRequest = URLRequest(url: URL(string: productsEndpoint)!)
        urlRequest.setValue("application/json", forHTTPHeaderField: "Content-Type")
        urlRequest.setValue("bearer \(accessToken)", forHTTPHeaderField: "Authorization")
        
        let task = URLSession.shared.dataTask(with: urlRequest) {(data, response, error) in
            let decoder = JSONDecoder()
            do {
                self.productsList = try decoder.decode([Product].self, from: data!)
                DispatchQueue.main.async {
                    self.tableView.reloadData()
                }
            } catch {
                print("error parsing json data")
                print(error)
            }
        }
        
        task.resume()
        
    }

    // MARK: - Table view data source

    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return productsList.count
    }

    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "productCell", for: indexPath)

        cell.textLabel?.text = productsList[indexPath.row].name
        
        let formatter = NumberFormatter()
        formatter.numberStyle = .currency
        if let formattedAmount = formatter.string(from: productsList[indexPath.row].price as NSNumber) {
            cell.detailTextLabel?.text  = formattedAmount
        }

        return cell
    }


}
